### PR TITLE
CLDR-15034 kbd: Keyboard 3.0 Spec

### PR DESCRIPTION
**Pull request by srl295**
- Update to the TR35 part 7 Keyboard spec, to harmonize with PRI 476 (techpreview)

Co-authored-by: Marc Durdin <marc@durdin.net>
(cherry picked from commit f3d2f68b2843a2e1939191a7ecd18ab974374d0a)

CLDR-15034

- [ ] This PR completes the ticket.

ALLOW_MANY_COMMITS=true

### CODE REVIEW

**Comment by srl295** [comment]
Review Note: this basically matches the PRI, I'd like to ask if we can merge this somewhat as-is and make further changes (editorial and otherwise) in new PRs on main.  Note that https://unicode-org.atlassian.net/browse/CLDR-16909 is for changing `techpreview` to `44`, also affecting the spec.

**Comment by miloush** [comment]
As we discussed in the group I would like to have a working implementation before we switch this from preview.

**Comment by srl295** [comment]
> As we discussed in the group I would like to have a working implementation before we switch this from preview.

Keyman is able to support a lot already- still working on regex implementation. So there is a lot that is implemented and also full validation is in the kmc compiler. I just got reordering working recently.

**Comment by mcdurdin** [comment]
I'd be comfortable with waiting for the implementation in Keyman to prove the spec before merging to main. I don't think there is any harm in waiting?

**Comment by srl295** [comment]
@miloush @mcdurdin merging here isn't final release, all i'm proposing here is to merge into in-development CLDR.  This way we go back to regular process for making PR changes. 

CLDR data slush is early August (2023), public alpha Aug 30, public beta Sep 20 , public beta Oct 18. 

1. "waiting for the implementation in Keyman" - do you mean a released implementation?  

2. "not removing techpreview" -  As I said I'm fine to leave "techpreview" on for the moment, perhaps until public beta timeframe. 

-----

To sum up,  what I would like to do with this PR and #3114 is to merge the PRI-level spec into CLDR main.

This would be followed up with PRs to address the feedback (which I am working on collating today).

**Comment by srl295** [comment]
ping… can i get a review on this? I'd like to merge this as-is and then address any changes in subsequent PRs.

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1631]: Given our recent discussions on markers in Keyman, I wonder if this advice may be unhelpful/misleading and perhaps we should just leave it to the implementers to figure out how they want to solve? PUA characters could be used by an experimental keyboard for a new orthography, even if that keyboard is not acceptable for publication into CLDR.

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1733]: This looks unfinished

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1751]: > UnicodeSets cannot be used directly in a transform.

This is slightly confusing -- perhaps we need to say something like:

```suggestion
    - Warning: Character classes look superficially similar to UnicodeSets as defined in [`unicodeSet`](#element-unicodeSet) elements, but they are different. UnicodeSets must be defined with a `unicodeSet` element, and referenced with the `$[unicodeSet]` notation in transforms. UnicodeSets cannot be specified inline in a transform, and can only be used indirectly by reference to the corresponding `unicodeSet` element.
```

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1770]: Non-capture groups?

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1871]: Implementation complexity?

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1903]: Is this right? not `$0`?

**Review by mcdurdin — APPROVED** [review]
This is a solid cleanup of the spec. I've got some minor nits and a couple of places which need further discussion (new tickets?), but nothing major that should block this from merging I think

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1903]: I think you are right. 


```suggestion
    `$0`
```

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:763]: Perhaps it should just say "does not participate until the next keystroke"

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1631]:
```suggestion
Markers should be implemented out-of-band from the normal text stream. Implementations should take care that all normal Unicode codepoints, including private-use area (PUA), are not affected by any scheme for encoding markers in text.
```

File: docs/ldml/tr35-keyboards.md
Comment (mcdurdin) [reply docs/ldml/tr35-keyboards.md:763]: Yep. And perhaps a note that using markers is another way to solve the problem?

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:713]: I'll remove the 'discouraged' and the link, there's other comments about rewriting the Accessibility section.

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:724]: I'll remove the 'discouraged' and the link, there's other comments about rewriting the Accessibility section.

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:722]: pre-existing.

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:763]: By the way, a marker that’s never directly consumed, whose purpose may be just to be not-a-character, might need to be noted as far as validation goes.

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1631]: This was implemented when this PR merged.

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1733]: Fixed in upcoming PR, noting that CLDR-16825 proposes removing \u1234 format

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:724]: * Addressed in upcoming PR

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:713]: * Addressed in upcoming PR

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:763]: adressed in upcoming PR

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1751]: Added to upcoming PR

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1770]: reworded in upcoming PR

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1871]: Yes good catch.!

Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1871]: added to upcoming PR

File: docs/ldml/tr35-keyboards.md
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1903]: added to upcoming PR

**Pull request merged**
